OpenRA.Game/Map/MapDirectoryTracker.cs

```diff
@@ -0,0 +1,135 @@
+using System.Collections.Generic;
+using System.IO;
+using System.Linq;
+using OpenRA.FileSystem;
+
+namespace OpenRA
+{
+	public sealed class MapDirectoryTracker : IDisposable
+	{
+		readonly FileSystemWatcher watcher;
+		readonly MapGrid mapGrid;
+		readonly IReadOnlyPackage package;
+		readonly MapClassification classification;
+
+		enum MapAction { Add, Delete, Update }
+		readonly Dictionary<string, MapAction> mapActionQueue = new Dictionary<string, MapAction>();
+
+		bool dirty = false;
+
+		public MapDirectoryTracker(MapGrid mapGrid, IReadOnlyPackage package, MapClassification classification)
+		{
+			this.mapGrid = mapGrid;
+			this.package = package;
+			this.classification = classification;
+
+			watcher = new FileSystemWatcher(package.Name);
+			watcher.Changed += (object sender, FileSystemEventArgs e) => AddMapAction(MapAction.Update, e.FullPath);
+			watcher.Created += (object sender, FileSystemEventArgs e) => AddMapAction(MapAction.Add, e.FullPath);
+			watcher.Deleted += (object sender, FileSystemEventArgs e) => AddMapAction(MapAction.Delete, e.FullPath);
+			watcher.Renamed += (object sender, RenamedEventArgs e) => AddMapAction(MapAction.Add, e.FullPath, e.OldFullPath);
+
+			watcher.IncludeSubdirectories = true;
+			watcher.EnableRaisingEvents = true;
+		}
+
+		public void Dispose()
+		{
+			watcher.Dispose();
+		}
+
+		void AddMapAction(MapAction mapAction, string fullpath, string oldFullPath = null)
+		{
+			lock (mapActionQueue)
+			{
+				dirty = true;
+
+				// if path is not root, update map instead
+				var path = RemoveSubDirs(fullpath);
+				if (fullpath == path)
+					mapActionQueue[path] = mapAction;
+				else
+					mapActionQueue[path] = MapAction.Update;
+
+				// called when file has been renamed / changed location
+				if (oldFullPath != null)
+				{
+					var oldpath = RemoveSubDirs(oldFullPath);
+					if (oldpath != null)
+						if (oldFullPath == oldpath)
+							mapActionQueue[oldpath] = MapAction.Delete;
+						else
+							mapActionQueue[oldpath] = MapAction.Update;
+				}
+			}
+		}
+
+		public void UpdateMaps(MapCache mapcache)
+		{
+			lock (mapActionQueue)
+			{
+				if (!dirty)
+					return;
+
+				dirty = false;
+				string lastUid = null;
+				foreach (var mapAction in mapActionQueue)
+				{
+					var map = mapcache.FirstOrDefault(x => x.Package?.Name == mapAction.Key && x.Status == MapStatus.Available);
+					if (map != null)
+					{
+						if (mapAction.Value == MapAction.Delete)
+						{
+							Console.WriteLine("Delete " + mapAction.Key);
+							map.Invalidate();
+						}
+						else
+						{
+							Console.WriteLine("Update " + mapAction.Key);
+							map.Invalidate();
+							lastUid = mapcache.LoadMap(mapAction.Key.Replace(package.Name + Path.DirectorySeparatorChar, ""), package, classification, mapGrid);
+
+							// Provide game lobby with information about our latest update
+							if (lastUid != null && map.Uid != lastUid)
+								mapcache.CallMapUpdated(map.Uid, lastUid);
```

Comment:
Whilst I'm happy if we only fire `MapUpdated` for updates and not add/deletes, the responsibility should still be on `MapCache` to choose when that happens rather than allowing any caller to trigger it via `CallMapUpdated`.

Could `LoadMap` have some logic to check if it's updating an existing entry rather than adding a new one, and fire the event?

Reply:
I moved `MapUpdated` to `LoadMap`. It is now also called when a map was added, letting me move `lastUpdatedUid` to `LobbyLogic`